### DISCUSSION TITLE
Docker Desktop error tpotinit is unhealthy (T-Pot 24.04)

### DISCUSSION THREAD

**Discussion by IKMAR00** [telekom-security/tpotce#1530] [Q&A] [answered]
Hello.

I am new to honeypots and my intention is to build a new honeypot for learning purposes.

I am trying to deploy the base project in a test environment (Windows 10 Home machine with Docker Desktop as the guide indicates).

Executing `genuser.sh` works perfectly but, when executing `docker compose up`, I keep getting this error caused by tpotinit:

![thumbnail_image](https://github.com/telekom-security/tpotce/assets/116032332/9efff0af-8ce7-4a2f-9aec-3ad47dd8d672)

In Docker Desktop I also have enabled WSL integration and Ubuntu.
![imagen](https://github.com/telekom-security/tpotce/assets/116032332/1f754467-86be-40fe-a56f-ce146f9d4d0c)
 
 Is there any step that I missed, or this is another type of issue?
 Thank you in advance!

**Comment by t3chn0m4g3** [comment]
Based on the info provided it looks like you have not copied the proper [compose file](https://github.com/telekom-security/tpotce?tab=readme-ov-file#macos--windows) and / or set T-Pot to run on Win in `.env`.

**Reply by t3chn0m4g3** [reply]
I could reproduce the behaviour.

**Reply by t3chn0m4g3** [reply]
This is fixed with #1531 
Run `update.sh` / or `git pull` to download the latest `compose/mac_win.yml` and replace `docker-compose.yml` with it.
If everything works fine a new `tpotinit` image will be pulled.
Please be aware that T-Pot on macOS and Windows is rather limited due to the constraints of Docker Desktop.
Thanks for taking the time to report this!

**Reply by IKMAR00** [reply]
Hi!
Thank you for your quick response!

I have repeated the whole process, erasing the old files and doing a `git pull`, but I still get the same error.

According to Docker Desktop, the new `tpotinit` image is being pulled because it indicates that the image was created 1 day ago.

I have followed the guide steps in order.
I have replaced the `docker-compose.yml` of the folder `~tpotce` with the new `mac_win.yml`
I can create a user without problems in my closed environment.
![imagen](https://github.com/telekom-security/tpotce/assets/116032332/f8433964-eb19-46e0-a716-52457b8b115c)

The .env was changed `TPOT_OSTYPE=linux` to `TPOT_OSTYPE=win` like the last time.

With the new tpotinit image, can you still reproduce the behavior?

**Reply by t3chn0m4g3** [reply]
Are you running in Docker Desktop for Windows, or are you using WSL? Based on the screenshot you seem to run in WSL or Linux.
We have never tested WSL only Docker Desktop for Windows. I would assume to use `TPOT_OSTYPE=linux` and the `standard.yml` in WSL. I do not even know if WSL supplies the necessary system resources to run T-Pot.
The Docker Desktop for Win / macOS runs perfectly fine now and at this point I cannot reproduce this.

**Reply by IKMAR00** [reply]
I was using the Ubuntu distro from the Microsoft Store because if I execute `genuser.sh` from PowerShell or cmd, it doesn’t do anything or shows an error.
Executing `genuser.sh` in PowerShell or cmd, it doesn't create the container and causes a pop-up window from Git for Windows to appear that shows the following text.
![imagen](https://github.com/telekom-security/tpotce/assets/116032332/c2f3ab2d-bc06-4381-8f2d-98fcd4e14069)

If I use PowerShell or cmd and put `bash genuser.sh`, it downloads `tpotinit` and the container is created, but I cannot set up a user.
![imagen](https://github.com/telekom-security/tpotce/assets/116032332/d59488a8-cb9f-4723-8a3b-9b11164f90a4)

I also have tried to just execute `docker run -v $HOME/tpotce:/data --entrypoint bash -it -u $(id -u):$(id -g) dtagdevsec/tpotinit:24.04 "/opt/tpot/bin/genuser.sh"` which also downloads the image and lets me create the user, but it indicates that ‘id’ is not recognized as the name of a cmdlet, function, script or executable.
![imagen](https://github.com/telekom-security/tpotce/assets/116032332/a4f3e945-7616-4cbf-b047-c0f2f2aebd6d)

I also have tried to remove the Ubuntu entirely and disable WSL from Windows features, but it ended up with Docker Desktop being unable to initialize until reinstalling Docker Desktop and reactivating WSL.

The other steps from the guide were followed after creating the user, but, it still indicates that `tpotinit` is unhealthy in Windows 10 Home.

**Reply by t3chn0m4g3** [reply]
What happens if you run the script like this:
```
#!/usr/bin/env bash
docker run -v $HOME/tpotce:/data --entrypoint bash -it -u 1000:1000 dtagdevsec/tpotinit:24.04 "/opt/tpot/bin/genuser.sh"
```

If that does not work either leave out `-u 1000:1000`.

And will `.env` actually reflect the changes for `WEB_USER`?

**Reply by IKMAR00** [reply]
Executing the updated `.\genuser.sh` causes the same effects has before, the pop-up window from Git for Windows with the same message.

However, executing `bash genuser.sh` doesn’t show the previous error; it creates the container, but doesn't update the `.env` file. Removing `-u 1000:1000` doesn’t make a difference in this step. 
![imagen](https://github.com/telekom-security/tpotce/assets/116032332/741ce449-8455-4a3f-8be3-e231759bc970)

I tried to manually input the user with the brackets it showed in the output console, but it showed the same error: `tpotinit is unhealthy`, and the output was three times instead of one.
![imagen](https://github.com/telekom-security/tpotce/assets/116032332/691925b5-2da3-4219-8ea7-ba68d7d43c81)

I have also tried to put the user without the brackets [] but it still showed that `tpotinit is unhealthy`.
![imagen](https://github.com/telekom-security/tpotce/assets/116032332/11f1991a-60ac-494d-b217-7029460cc610)

**Reply by t3chn0m4g3** [reply]
Thanks. I could finally reproduce this with a complete fresh Win 11 / Docker Desktop install. It seems Docker Desktop switched to WSL2 internally which shows up the Docker VM `uname -a` actually as `linux` instead of `linuxkit` which in turn leaves `entrypoint.sh` with improper checks.
Switching back to WSL will not help, because `docker compose` handles volumes differently on Windows, now all volume file handles must be present at the project start, instead of the individual container which results in `nginxpasswd` always missing.
I have a working prototype now, but this still needs refinement and testing with Linux, macOS versions before being pushed.

![image](https://github.com/telekom-security/tpotce/assets/4318452/c578d752-2478-42ab-ab65-621743a35eec)

**Reply by IKMAR00** [reply]
Oh, good!

I'll be waiting for the new version when it's ready.

Thank you so much for everything!

**Reply by t3chn0m4g3** [reply, marked as answer]
I just pushed the [changes](https://github.com/telekom-security/tpotce/commit/0b1281d40fa34233903be9d739833654eeb34256).
At best you run a `git reset --hard` and then a `git pull` and follow the [ReadMe](https://github.com/telekom-security/tpotce/blob/master/README.md#macos--windows).

**Reply by IKMAR00** [reply]
Yes!
It's working!
 
![imagen](https://github.com/telekom-security/tpotce/assets/116032332/2e230132-a953-4149-8d07-17b83db9e73c)

I also noticed that Ciscoasa gets triggered if you execute `docker compose up` after stopping T-Pot with `CTRL-C`. 
It has to be a false positive since, by default, Windows firewall blocks access from remote as the ReadMe indicates.

No other honeypots get triggered except Ciscoasa, and, if I am not mistaken, setting up T-Pot in an unfiltered zone should register attacks from multiple honeypots in a few minutes.

Thank you for everything!
